peps/pep-0789.rst

```diff
@@ -0,0 +1,468 @@
+PEP: 789
+Title: Preventing task-cancellation bugs by limiting yield in async generators
+Author: Zac Hatfield-Dodds <zac@zhd.dev>,
+        Nathaniel J. Smith <njs@pobox.com>
+PEP-Delegate:
+Discussions-To: https://discuss.python.org/t/preventing-yield-inside-certain-context-managers/1091
+Status: Draft
+Type: Standards Track
+Created: 14-May-2024
+Python-Version: 3.14
+
+
+Abstract
+========
+
+`Structured concurrency`_ is increasingly popular in Python.  Interfaces such as
+the ``asyncio.TaskGroup`` and ``asyncio.timeout`` context managers support
+compositional reasoning, and allow developers to clearly scope the lifetimes of
+concurrent tasks. However, using ``yield`` to suspend a frame inside such a
+context (a "CancelScope") corrupts this structure, resulting in situations where
+the wrong task is canceled, timeouts are not respected, and exceptions are not
+properly propagated or handled.
+
+To address this issue, this PEP proposes a new ``sys.block_yields()`` context
+manager. When inside this context, attempting to ``yield`` will raise a
+RuntimeError, preventing the task from yielding across a scope boundary.
+Additionally, a mechanism will be provided to allow generator-based context
+managers to permit yields across a single call.  ``sys.block_yields()`` will be
+used by asyncio and downstream libraries to implement task groups, timeouts, and
+cancellation; and by ``contextlib`` etc. to convert generators into context
+managers which allow safe yields.
+
+.. _Structured concurrency: https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/
+
+Motivation
+==========
+
+Structured concurrency is increasingly popular in Python, in the form of newer
+``asyncio`` interfaces and third-party libraries such as Trio and anyio.  These
+interfaces support compositional reasoning, *so long as* users never write a
+``yield`` which takes control-flow across a ``CancelScope``.
+
+A ``CancelScope`` is a context manager which can... cancel... whatever work occurs
```

Comment:
Since you're using this as a generic term rather than a reference to a specific object, may be better to write "cancel scope".